File: part1/src/graph.ts
import { Annotation, StateGraph, START, END } from '@langchain/langgraph';
import { generateMessage } from './llm';

interface CoffeeCustomizations {
    creams: number,
    milks: number,
    sugars: number,
    sweeteners: number,
    whippedCream: boolean,
}

const CoffeeAnnotation = Annotation.Root({
    customerName: Annotation<string>,
    coffeeName: Annotation<string>,
    customizations: Annotation<CoffeeCustomizations>,
    message: Annotation<string>,
});

const stepGenerateMessage = async (state: typeof CoffeeAnnotation.State) => {
    const coffeeOptions = {
        coffeeName: state.coffeeName,
        customizations: state.customizations
    }

    // Call the LLM
    const message = await generateMessage(state.customerName, coffeeOptions)
    state.message = message;

    // Return the state
    return state;
}

const coffeeMessageGraph = new StateGraph(CoffeeAnnotation)
    .addNode("generateMessage", stepGenerateMessage)
    .addEdge(START, "generateMessage")
    .addEdge("generateMessage", END)
    .compile();

export const runCoffeeMessageGraph = async (initialInput: typeof CoffeeAnnotation.State) => {
    return coffeeMessageGraph.invoke(initialInput);
}

File: pythonversion/part1/src/graph.py
from __future__ import annotations

from typing import TypedDict, Any, Dict
from langgraph.graph import StateGraph, START, END
from .llm import generate_message


class CoffeeCustomizations(TypedDict, total=False):
    creams: int
    milk: int
    sugars: int
    sweeteners: int
    whippedCream: bool


class CoffeeState(TypedDict, total=False):
    customerName: str
    coffeeName: str
    customizations: CoffeeCustomizations
    message: str


def _step_generate_message(state: CoffeeState) -> CoffeeState:
    coffee_options = {
        "coffeeName": state.get("coffeeName"),
        "customizations": state.get("customizations", {}),
    }
    msg = generate_message(state.get("customerName", "Friend"), coffee_options)
    state["message"] = msg
    return state


_graph = StateGraph(CoffeeState)
_graph.add_node("generateMessage", _step_generate_message)
_graph.add_edge(START, "generateMessage")
_graph.add_edge("generateMessage", END)
_compiled = _graph.compile()


def run_coffee_message_graph(initial_input: Dict[str, Any]) -> CoffeeState:
    return _compiled.invoke(initial_input)

